src/Microsoft.TemplateEngine.Cli/TemplatePackageCoordinator.cs

```diff
@@ -65,6 +66,54 @@ internal class TemplatePackageCoordinator
             return (await managedTemplatePackage.ManagedProvider.GetLatestVersionsAsync(new[] { managedTemplatePackage }, cancellationToken).ConfigureAwait(false)).Single();
         }
 
+        internal async Task<(string, string)> ValidatePackageAvailabilityAsync(ITemplateInfo template, CancellationToken cancellationToken)
+        {
+            _ = template ?? throw new ArgumentNullException(nameof(template));
+            cancellationToken.ThrowIfCancellationRequested();
+
+            ITemplatePackage templatePackage;
+            try
+            {
+                templatePackage = await _templatePackageManager.GetTemplatePackageAsync(template, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception)
+            {
+                Reporter.Error.WriteLine(string.Format(LocalizableStrings.TemplatePackageCoordinator_Error_PackageForTemplateNotFound, template.Identity));
+                return default;
+            }
+
+            if (!(templatePackage is IManagedTemplatePackage managedTemplatePackage))
+            {
+                //update is not supported - built-in or optional workload source
+                return default;
+            }
+
+            var templatePackages = await _templatePackageManager.GetTemplatePackagesAsync(force: false, cancellationToken).ConfigureAwait(false);
+            IReadOnlyList<(string Id, string Version)> unmanagedTemplatePackages = templatePackages
+                .Where(tp => tp is not IManagedTemplatePackage)
+                .Select(tp => NuGetUtils.GetNuGetPackageInfo(_engineEnvironmentSettings, tp.MountPointUri))
+                .Where(i => i != default)
+                .ToList();
```

Comment:
We are interested only in a first match - so it is better to keep this as `IEnumarable`

Reply:
fixed